### PR TITLE
Allow adding code to EthAccount in InitGenesis

### PR DESCRIPTION
**Pull request by itsdevbear**
Closes: #[921](https://github.com/tharsis/ethermint/issues/921)

## Description

Allow adding code to EthAccounts at Genesis

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/evm/genesis.go [hunk 1]
```diff
@@ -51,12 +52,21 @@ func InitGenesis(
 
 		code := common.Hex2Bytes(account.Code)
 		codeHash := crypto.Keccak256Hash(code)
-		if !bytes.Equal(ethAcct.GetCodeHash().Bytes(), codeHash.Bytes()) {
-			panic("code don't match codeHash")
+		isEmptyCodeHash := codeHash.Hex() == common.BytesToHash(types.EmptyCodeHash).String()
```
Comment (fedekunze) [inline x/evm/genesis.go:55]: use `bytes.Equal` to avoid unnecessary hex conversion

File: x/evm/genesis.go [hunk 2]
```diff
@@ -51,12 +52,21 @@ func InitGenesis(
...
+
+		// If account has no code allow for code to be added, if code already exists do not allow modification
+		if !isEmptyCodeHash && bytes.Equal(ethAccount.GetCodeHash().Bytes(), codeHash.Bytes()) {
+			panic("Code don't match codeHash")
```
Comment (fedekunze) [inline x/evm/genesis.go:59]: can you add an additional unit test for this case? (see `genesis_test.go`)

**Review by fedekunze — COMMENTED** [review]
Code LGTM. Minor comments

Wondering why you didn't prefer using a precompiled contract or setting the contract code hash to the account as well?

File: x/evm/genesis.go (hunk 1 above)
Comment (yihuang) [reply x/evm/genesis.go:55]: it should be trivial to update the code hash to the correct value in the genesis prepare script. I'm not sure if it's worth trading validation to convenience to manually modifying the genesis file.

Comment (itsdevbear) [reply x/evm/genesis.go:55]: Hey guys, sorry this totally fell off my radar, if the codehash doesn't match I believe it will cause problems down the road, so a sanity check at genesis is probably not the worst idea in the world.

**Comment by itsdevbear** [comment]
@fedekunze made the changes.

For my workflow I wrote a script that compiles my contracts via forge foundry, and then inserts the bytecode directly into genesis.json.

So my init.sh effectively adds
````
cd contracts
forge build
cd ..
// take bytecode, and insert into genesis.json
```

The other thing we could do is create a more formal way of doing this right into the golang code. Like how you do it in Evmos for the ERC20 module. (`contracts.ERC20Decimals....`). Then in init genesis of the evm module, check to see what compiled contract golang objects exist and slot them in. The only thing is that this would mess with exporting stuff, so I think the bytecode needs to be in genesis.json for sanity check reasons. 

TLDR: I wanted a way to be able to insert contracts easily at genesis while without having to do anything manually. When working and developing genesis contracts I found it a pain to have to go and manually deploy contracts after booting things up, makes automated testing difficult.

**Comment by yihuang** [comment]
> @fedekunze made the changes.
> 
> For my workflow I wrote a script that compiles my contracts via forge foundry, and then inserts the bytecode directly into genesis.json.
> 
> So my init.sh effectively adds
> 
> ```
> cd contracts
> forge build
> cd ..
> // take bytecode, and insert into genesis.json
> ```
> 
> The other thing we could do is create a more formal way of doing this right into the golang code. Like how you do it in Evmos for the ERC20 module. (`contracts.ERC20Decimals....`). Then in init genesis of the evm module, check to see what compiled contract golang objects exist and slot them in. The only thing is that this would mess with exporting stuff, so I think the bytecode needs to be in genesis.json for sanity check reasons.
> 
> TLDR: I wanted a way to be able to insert contracts easily at genesis while without having to do anything manually. When working and developing genesis contracts I found it a pain to have to go and manually deploy contracts after booting things up, makes automated testing difficult.

we can write a command or script to update code and codeHash at the same time?

**Comment by itsdevbear** [comment]
@yihuang so provide codeHash at genesis opposed to generate it in initGenesis? Whats your thought process there? Just seems like additional work to be done outside of the core codebase. Thoughts?

**Comment by yihuang** [comment]
> @yihuang so provide codeHash at genesis opposed to generate it in initGenesis? Whats your thought process there? Just seems like additional work to be done outside of the core codebase. Thoughts?

yeah, I think it keeps the genesis file canonical.

**Comment by itsdevbear** [comment]
@yihuang , I added the codehash to the genesis file.

**Comment by itsdevbear** [comment]
Looks like some of the proto files were regenerated and created an ugly diff, I can revert if needed (I just ran `starport generate proto-go`)

File: proto/ethermint/evm/v1/genesis.proto [hunk 3]
```diff
@@ -22,7 +22,9 @@ message GenesisAccount {
   string address = 1;
   // code defines the hex bytes of the account code.
   string code = 2;
+  // codeHash defines the keccak256 hash of the hex bytes of the account code.
+  string codeHash = 3;
```
Comment (yihuang) [inline proto/ethermint/evm/v1/genesis.proto:26]: I don't get it, isn't it always `keccak256(code)`, maybe I have some misunderstanding here.

Comment (itsdevbear) [reply proto/ethermint/evm/v1/genesis.proto:26]: That codeHash will not be set at this point in time. It does not get set at genesis currently.

Comment (yihuang) [reply proto/ethermint/evm/v1/genesis.proto:26]: when you export a genesis, the codeHash under the auth section should be correct.
I don't get what's the issue we are trying to solve here?

Comment (itsdevbear) [reply proto/ethermint/evm/v1/genesis.proto:26]: Motivation for this PR was to add precompiled contracts to the chain before the first block is produced.

Comment (yihuang) [reply proto/ethermint/evm/v1/genesis.proto:26]: I see, I've done sth like this before, what I do is start a local devnet, deploy the contract, and export the genesis file.
It should be possible to directly modify the genesis file too, what's the obstacle of doing that right now?

Comment (yihuang) [reply proto/ethermint/evm/v1/genesis.proto:26]: https://github.com/crypto-org-chain/cronos/blob/main/integration_tests/configs/genesis_token_mapping.yaml#L44

Comment (itsdevbear) [reply proto/ethermint/evm/v1/genesis.proto:26]: image
